### DISCUSSION TITLE
Saving 12-bit tiff images

### DISCUSSION THREAD

**Discussion by stiefel512** [ImageMagick/ImageMagick#4525] [Help]
I have a grayscale tiff where all values are in the range 0-4095. I would like to pack the bits, but keep the pixel values fixed.
`convert <input.tiff> -depth 12  <output.tiff>` packs the bits, but changes the pixel values. What am I missing?

Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101
Ubuntu 20.04.3 LTS

**Comment by snibgo** [comment]
Please link to a sample input.

**Reply by stiefel512** [reply]
I put a sample input and output here:
https://drive.google.com/drive/folders/1PQGuIzer4kuhhatlLoHHQ9iz0w9U3Dmh?usp=sharing

**Comment by snibgo** [comment]
test_input.tiff has one channel, 16-bit integer. The maximum value is 4094, so only the _bottom_ 12 bits are used.

"-depth 12" saves the image using (speaking roughly) the _top_ 12 bits. To get what you want with Q16 IM, multiply by (roughly) 16. 
```
convert test_input.tiff -evaluate Multiply 16.003663003663 -depth 12 x.tiff
```
Note that (2^16-1) / (2^12-1) = 16.003663003663

EDIT: Added "-depth 12".

**Reply by stiefel512** [reply]
The result is off by 2 in every non-0 value. I suspect this has something to do with the storage type- my pixel values are unsigned shorts (or uint16). The maximum pixel value in 12 bits is 4095. However, I have not succeeded in getting the correct values

`convert test_input.tiff -evaluate Multiply 16 x.tiff` does give the correct multiplication, so it's just in the depth step where the function gets lost.

As an aside, moving the bits from the bottom to the top bits of the register is a bit shift. It can be done either by multiplying by 2^n (where n is the number of bits to shift) or by performing a LeftShift n
I tried using -evaluate LeftShift 4, which should do the same as Multiply 16, and I get the same result

**Reply by snibgo** [reply]
> The result is off by 2 in every non-0 value.

Oops, sorry. I haven't investigated, but I may have been wrong.

In general, when changing depth, IM ensures the maximum value in one depth transforms to the maximum value in the other bit depth, so it multiplies by the ratio of the maximum values. The maximum value at depth Q is (2^Q-1). For example, when promoting from 8 bits to 16, it multiplies by (2^16-1)/(2^8-1).

I thought that is what it would do with 12 bits, but I could be wrong. Perhaps it only applies when bit depths are multiples of 8.

**Reply by stiefel512** [reply]
I suspect that the issue may be related to signed/unsigned values, but I am not sure how to test or handle this. The pixels are unsigned integers- is there a way to know how convert is reading those values?

**Comment by stiefel512** [comment]
Awesome. Thanks
